### PR TITLE
feat(796): oci artifact spec

### PR DESCRIPTION
**Pull request by matthiasbruns**
#### What this PR does / why we need it
First PR to add oci artifact transformers in the bindings lib.
Part of https://github.com/open-component-model/open-component-model/pull/1684

GetOCIArtifactSpec will download a artifact and put it into a file defined by `OutputPath` or into tmp when `OutputPath` is empty.
The next step is to integrate the spec into the cli `transfer component-version`.

```go
type GetOCIArtifactSpec struct {
	// Resource is the resource descriptor to get the OCI artifact from.
	Resource *v2.Resource `json:"resource"`
	// OutputPath is the path where the artifact should be downloaded to..
	// If empty, a temporary file will be created.
	OutputPath string `json:"outputPath,omitempty"`
}
```

#### Which issue(s) this PR fixes
Contributes: https://github.com/open-component-model/ocm-project/issues/796

### CODE REVIEW

File: bindings/go/oci/integration/integration_test.go [hunk 1]
```diff
@@ -60,6 +67,18 @@ const (
...
+	Password string
+}
+
+func (m mockCredentialsResolver) Resolve(ctx context.Context, identity ocmruntime.Identity) (map[string]string, error) {
+	return map[string]string{
+		"username": m.Username,
+		"password": m.Password,
+	}, nil
+}
+
```
Comment (fabianburth) [inline bindings/go/oci/integration/integration_test.go]: whats the value add of this mock cred resolver, are you testing against it in any way? couldnt you just pass nil?

File: bindings/go/oci/transformer/get_oci_artifact.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+		}
+	}
+
+	blobContent, err := t.Repository.DownloadResource(ctx, targetResource, creds)
+	if err != nil {
+		return nil, fmt.Errorf("failed downloading OCI artifact %v %w", resource.ToIdentity(), err)
+	}
+
+	// Determine output path
+	if outputPath == "" {
```
Comment (jakobmoellerdev) [inline bindings/go/oci/transformer/get_oci_artifact.go]: youre duplicating code here I believe, best to use a shared method maybe?

**Review by jakobmoellerdev — COMMENTED** [review]
mostly lgtm, just smaller corrections and questions I believe

File: bindings/go/oci/integration/integration_test.go [hunk 3]
```diff
@@ -60,6 +67,18 @@ const (
 	userAgent                 = "ocm.software"
 )
 
+type mockCredentialsResolver struct {
```
[the remark this answers is not in the document]
Comment (matthiasbruns) [reply bindings/go/oci/integration/integration_test.go]: Maybe you understood me wrong - I do not feel that we need this helper in production code just to have a static credentials helper for testing - I will add this now, but again, I do not feel that there is something broken that needs to be fixed

[the remark this answers is not in the document]
Comment (matthiasbruns) [reply bindings/go/oci/integration/integration_test.go]: https://github.com/open-component-model/open-component-model/pull/1736
I am not sure if I should have used the DAG for this or added it into the existing resolver flow from the dag

[the remark this answers is not in the document]
Comment (jakobmoellerdev) [reply bindings/go/oci/integration/integration_test.go]: This resolver you wrote is totally fine! I think you wont be the last person to need static credentials without wanting a full credential graph, so having this in the binding library is useful either way.

File: bindings/go/oci/integration/integration_test.go [hunk 4]
```diff
@@ -285,6 +304,14 @@ func Test_Integration_OCIRepository(t *testing.T) {
 		t.Run("oci image digest processing", func(t *testing.T) {
 			processResourceDigest(t, repo, "ghcr.io/test:v1.0.0", reference("new-test:v1.0.0"))
 		})
+
+		t.Run("get oci artifact", func(t *testing.T) {
```
Comment (jakobmoellerdev) [inline bindings/go/oci/integration/integration_test.go]: this should not be part of the backwards compatibility tests. also it seems like you structure the tests a lot differently than the others. Why is that? I was expecting you to setup a new test set just for the transformers here

Comment (matthiasbruns) [reply bindings/go/oci/integration/integration_test.go]: So I had a second look why I did the setup

The existing integration tests in oci relied on either `repository.ComponentVersionRepository` or `oci.Repository`

The transformer that is implemented here does not - it uses `repository.ResourceRepository`. I cannot use the  `oci.Repository`directly with the resolver setup since we did not change its api to support credential injection during up/download. This is why I had to setup the tests differently and use the `credentials.Resolver` instead the `urlresolver.Resolver` path.
I moved the transformer tests into their own parent tests case that we can add more tests there in the future.

Comment (jakobmoellerdev) [reply bindings/go/oci/integration/integration_test.go]: Makes sense 👍 I think all of these findings are valid. Ideally you will need to extract such follow ups into discussions on the warroom or at least point them out with a godoc comment on the test (rather the second as I dont think any of this is particularly important)

Comment (matthiasbruns) [reply bindings/go/oci/integration/integration_test.go]: docs added :)

File: bindings/go/oci/spec/transformation/v1alpha1/get_oci_artifact.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+// GetOCIArtifactSpec is the input specification for the
+// GetOCIArtifact transformation.
+// +k8s:deepcopy-gen=true
+// +ocm:jsonschema-gen=true
+type GetOCIArtifactSpec struct {
+	// Resource is the resource descriptor to get the OCI artifact from.
+	Resource *v2.Resource `json:"resource"`
+	// OutputPath is the path where the artifact should be downloaded to..
```
Comment (fabianburth) [inline bindings/go/oci/spec/transformation/v1alpha1/get_oci_artifact.go]:
```suggestion
	// OutputPath is the path where the artifact should be downloaded to.
```

**Review by Skarlso — COMMENTED** [review]
Generally okay, just a few little questions. :)
